### ISSUE TITLE
Are gradients summed or averaged across cores?

### ISSUE THREAD

**Issue by hrbigelow** [pytorch/xla#1466]
## 📚 Documentation

Hi,

I was looking at the docs but couldn't find an answer to this.  I was just wondering what the multi-core TPU workflow is?  As I understand it, at each training step, each TPU core contains a separate, identical copy of the model.  Then, the ParallelLoader loads a separate batch of data for each core.  The forward() and backward() calls are executed, producing 8 sets of gradients.

Then, what happens?  Is it that the average gradients across cores is computed, and an optimization step is taken on that average?  Or, is it the sum?

I assume then that an optimizer step is taken, yielding a new set of weights, and the weights are broadcasted to each of the cores, updating the copies of the model, is that right?

Also, if I wanted to do the same training on a GPU, would I simply use a batch 8 times as large, and get the same results, assuming the same random seeds and order of data that is yielded from the iterator?

Thanks

**Comment by dlibenzi** [comment]
What happens when you call `xm.optimizer_step()` is that we go in all the tensors which represents the gradients, and inject an `AllReduce(SUM)` IR node into their IR graphs, followed by the scaling by `1.0 / NUM_REPLICAS`, and then call the user's `optimizer.step()`.
When the PyTorch optimizer does its work in applying gradients to weight, it uses the averaged gradients.
On the TPU the XLA compiler generates very fast AllReduce operations over a dedicated link which connects the TPU chips.

Currently using PyTorch/XLA via the XLA GPU support, works, but the gradient reduction does not.
It is also not much tested.

In our code when we say "batch" we usually mean "mini batch" which stands for "per core batch".
On GPU I guess it depends on the model implementation whether the batch is the global batch or the mini batch size.

**Comment by hrbigelow** [comment]
Great, thanks so much for the explanation @dlibenzi - that is what I hoped and was expecting.  Yes, to clarify, when I used the word "batch", I meant "mini-batch" in the SGD sense.
